File: challenges/linked-lists/reverse-linked-list/builder/builder.py
import ast
import importlib.util
import json
import pathlib
import sys


def parse_value(raw: str):
    value = raw.strip()
    if value == "":
        return ""
    lowered = value.lower()
    if lowered == "null":
        return None
    if lowered == "true":
        return True
    if lowered == "false":
        return False
    if value.startswith("[") or value.startswith("{"):
        normalized = value.replace("null", "None").replace("true", "True").replace("false", "False")
        return ast.literal_eval(normalized)
    if (value.startswith('"') and value.endswith('"')) or (value.startswith("'") and value.endswith("'")):
        return value[1:-1]
    try:
        return int(value)
    except ValueError:
        pass
    try:
        return float(value)
    except ValueError:
        pass
    return value


def split_key_value(line: str):
    key, value = line.split(":", 1)
    return key.strip(), value.strip()


def parse_tests(path: pathlib.Path):
    lines = path.read_text(encoding="utf-8").splitlines()
    tests = []
    section = "visible"
    i = 0
    while i < len(lines):
        line = lines[i].rstrip("\n")
        stripped = line.strip()
        if not stripped:
            i += 1
            continue
        if stripped in ("visible:", "hidden:"):
            section = stripped[:-1]
            i += 1
            continue
        if line.startswith("  - "):
            case = {}
            first = line[4:].strip()
            i += 1
            if first:
                key, value = split_key_value(first)
                case[key] = parse_value(value)
            while i < len(lines):
                nxt = lines[i].rstrip("\n")
                nxt_stripped = nxt.strip()
                if nxt.startswith("  - ") or nxt_stripped in ("visible:", "hidden:"):
                    break
                if not nxt_stripped:
                    i += 1
                    continue
                indent = len(nxt) - len(nxt.lstrip(" "))
                if indent >= 4 and ":" in nxt:
                    key, value = split_key_value(nxt[indent:].strip())
                    case[key] = parse_value(value)
                i += 1
            case["visible"] = section == "visible"
            tests.append(case)
            continue
        i += 1
    return tests


def parse_challenge(path: pathlib.Path):
    meta = {}
    for line in path.read_text(encoding="utf-8").splitlines():
        stripped = line.strip()
        if not stripped or stripped.startswith("#"):
            continue
        if line.startswith(" "):
            continue
        if ":" not in stripped:
            continue
        key, value = split_key_value(stripped)
        meta[key] = value.strip('"')
    return meta


def load_module(solution_path: pathlib.Path):
    spec = importlib.util.spec_from_file_location("user_solution", solution_path)
    if spec is None or spec.loader is None:
        raise RuntimeError("unable to load solution")
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module


def resolve_function(module):
    for name in ("reverse_list", "reverseList", "ReverseList"):
        fn = getattr(module, name, None)
        if callable(fn):
            return fn
    raise RuntimeError("missing reverse_list function")


def resolve_list_node(module):
    cls = getattr(module, "ListNode", None)
    if cls is None:
        class ListNode:
            def __init__(self, val=0, next=None):
                self.val = val
                self.next = next
        return ListNode
    return cls


def array_to_linked_list(values, node_cls):
    head = None
    tail = None
    for value in values:
        node = node_cls(value)
        if head is None:
            head = node
            tail = node
            continue
        tail.next = node
        tail = node
    return head


def linked_list_to_array(head):
    values = []
    seen = 0
    node = head
    while node is not None and seen < 100000:
        values.append(node.val)
        node = node.next
        seen += 1
    return values


def render(value):
    if isinstance(value, (list, dict, tuple, int, float, bool)) or value is None:
        return json.dumps(value, separators=(",", ":"))
    return str(value)


def run(solution_path: pathlib.Path, tests_path: pathlib.Path, challenge_path: pathlib.Path):
    module = load_module(solution_path)
    fn = resolve_function(module)
    node_cls = resolve_list_node(module)
    tests = parse_tests(tests_path)
    meta = parse_challenge(challenge_path)
    input_type = meta.get("input_type", "array")
    output_type = meta.get("output_type", "array")

    failures = 0
    for idx, test in enumerate(tests, start=1):
        raw_input = test.get("input", [])
        if input_type == "linked-list":
            call_input = array_to_linked_list(raw_input, node_cls)
        else:
            call_input = raw_input

        got = fn(call_input)
        if output_type == "linked-list":
            got_value = linked_list_to_array(got)
        else:
            got_value = got

        expected = test.get("expected")
        if got_value == expected:
            print(f"PASS {idx}")
        else:
            failures += 1
            print(f"FAIL {idx} got={render(got_value)} expected={render(expected)}")
    return failures == 0


def main():
    builder_dir = pathlib.Path(__file__).resolve().parent
    solution_path = pathlib.Path(sys.argv[1]).resolve() if len(sys.argv) > 1 else (builder_dir.parent / "setup" / "python.py")
    tests_path = pathlib.Path(sys.argv[2]).resolve() if len(sys.argv) > 2 else (builder_dir.parent / "tests.yaml")
    challenge_path = pathlib.Path(sys.argv[3]).resolve() if len(sys.argv) > 3 else (builder_dir.parent / "challenge.yaml")
    ok = run(solution_path, tests_path, challenge_path)
    raise SystemExit(0 if ok else 1)


if __name__ == "__main__":
    main()

File: challenges/linked-lists/reverse-linked-list/builder/builder.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type testCase struct {
	Input    []int
	Expected []int
}

func main() {
	testsPath := "../tests.yaml"
	challengePath := "../challenge.yaml"
	if len(os.Args) > 2 {
		testsPath = os.Args[2]
	}
	if len(os.Args) > 3 {
		challengePath = os.Args[3]
	}

	tests, err := parseTests(testsPath)
	if err != nil {
		fmt.Fprintln(os.Stderr, err.Error())
		os.Exit(1)
	}
	inputType, outputType, err := parseTypes(challengePath)
	if err != nil {
		fmt.Fprintln(os.Stderr, err.Error())
		os.Exit(1)
	}

	failed := false
	for i, tc := range tests {
		var got []int
		if inputType == "linked-list" {
			head := sliceToList(tc.Input)
			out := reverseList(head)
			if outputType == "linked-list" {
				got = listToSlice(out)
			}
		}
		if outputType != "linked-list" {
			got = tc.Input
		}

		if equalSlices(got, tc.Expected) {
			fmt.Printf("PASS %d\n", i+1)
			continue
		}
		failed = true
		fmt.Printf("FAIL %d got=%s expected=%s\n", i+1, formatSlice(got), formatSlice(tc.Expected))
	}

	if failed {
		os.Exit(1)
	}
}

func parseTypes(path string) (string, string, error) {
	file, err := os.Open(path)
	if err != nil {
		return "array", "array", err
	}
	defer file.Close()

	inputType := "array"
	outputType := "array"
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		if strings.HasPrefix(line, " ") {
			continue
		}
		trimmed := strings.TrimSpace(line)
		if !strings.Contains(trimmed, ":") {
			continue
		}
		key, value := splitKV(trimmed)
		value = strings.Trim(value, "\"'")
		if key == "input_type" {
			inputType = value
		}
		if key == "output_type" {
			outputType = value
		}
	}
	if err := scanner.Err(); err != nil {
		return "array", "array", err
	}
	return inputType, outputType, nil
}

func parseTests(path string) ([]testCase, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var tests []testCase
	var current *testCase
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		trimmed := strings.TrimSpace(line)
		if trimmed == "" || trimmed == "visible:" || trimmed == "hidden:" {
			continue
		}
		if strings.HasPrefix(trimmed, "- ") {
			if current != nil {
				tests = append(tests, *current)
			}
			current = &testCase{}
			trimmed = strings.TrimSpace(strings.TrimPrefix(trimmed, "- "))
		}
		if current == nil || !strings.Contains(trimmed, ":") {
			continue
		}
		key, value := splitKV(trimmed)
		switch key {
		case "input":
			vals, err := parseIntSlice(value)
			if err != nil {
				return nil, err
			}
			current.Input = vals
		case "expected":
			vals, err := parseIntSlice(value)
			if err != nil {
				return nil, err
			}
			current.Expected = vals
		}
	}
	if err := scanner.Err(); err != nil {
		return nil, err
	}
	if current != nil {
		tests = append(tests, *current)
	}
	return tests, nil
}

func splitKV(line string) (string, string) {
	parts := strings.SplitN(line, ":", 2)
	if len(parts) == 1 {
		return strings.TrimSpace(parts[0]), ""
	}
	return strings.TrimSpace(parts[0]), strings.TrimSpace(parts[1])
}

func parseIntSlice(value string) ([]int, error) {
	raw := strings.TrimSpace(value)
	raw = strings.TrimPrefix(raw, "[")
	raw = strings.TrimSuffix(raw, "]")
	raw = strings.TrimSpace(raw)
	if raw == "" {
		return []int{}, nil
	}
	parts := strings.Split(raw, ",")
	result := make([]int, 0, len(parts))
	for _, part := range parts {
		n, err := strconv.Atoi(strings.TrimSpace(part))
		if err != nil {
			return nil, err
		}
		result = append(result, n)
	}
	return result, nil
}

func sliceToList(values []int) *ListNode {
	var head *ListNode
	var tail *ListNode
	for _, v := range values {
		n := &ListNode{Val: v}
		if head == nil {
			head = n
			tail = n
			continue
		}
		tail.Next = n
		tail = n
	}
	return head
}

func listToSlice(head *ListNode) []int {
	result := []int{}
	for node := head; node != nil; node = node.Next {
		result = append(result, node.Val)
	}
	return result
}

func equalSlices(a, b []int) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

func formatSlice(values []int) string {
	if len(values) == 0 {
		return "[]"
	}
	parts := make([]string, 0, len(values))
	for _, v := range values {
		parts = append(parts, strconv.Itoa(v))
	}
	return "[" + strings.Join(parts, ",") + "]"
}
